Transition scenarios

gwenchee opened the pull request:
This PR contains a performance test that gives a basic transition scenario from ten initial reactors to a fleet of new reactors where there is a constant, growing or sinusoidal power demand. There are source, reactor and sink facilities in this simulation. The source and reactor facilities are deployed using DemandDrivenDeploymentInst and the sink facilities are deployed using SupplyDrivenDeploymentInst. 

In `tests/performance_tests`, by running `python algorithm_performance_tests_transitions.py [transitiontype]`, the plots shown below will be generated, including the plots for other commodities (fuel, spentfuel)

Examples of the power commodity plots for each type of scenario: 
![constanttransition-power](https://user-images.githubusercontent.com/29397884/58265605-9d7de600-7d45-11e9-8c26-1419a6c26c35.png)
![growingtransition-power](https://user-images.githubusercontent.com/29397884/58265610-a078d680-7d45-11e9-99a4-2668a150498b.png)
![sinetransition-power](https://user-images.githubusercontent.com/29397884/58265615-a2429a00-7d45-11e9-92a0-c876a0e47211.png)

gwenchee committed:
Merge pull request #106 from arfc/master

update

gwenchee committed:
constant transition

gwenchee committed:
Merge pull request #107 from gwenchee/capz

Capz

gwenchee committed:
set ylim 0 and legend lower

gwenchee committed:
pep8

gwenchee committed:
color map

gwenchee committed:
growing transition

gwenchee committed:
better colormap

gwenchee committed:
spent fuel tracking

gwenchee committed:
const transition back deploy sink

gwenchee committed:
sine

gwenchee committed:
Merge branch 'master' into constanttransition

gwenchee committed:
pep8

gwenchee committed:
pep8

gwenchee committed:
pep8 sucks

gwenchee committed:
pep8

gwenchee committed:
Merge pull request #109 from gwenchee/constanttransition

update

gwenchee committed:
Merge pull request #111 from gwenchee/constanttransition

update

gwenchee committed:
Merge branch 'growingtransition' into constanttransition

gwenchee committed:
Merge pull request #112 from gwenchee/constanttransition

Constanttransition

gwenchee committed:
Merge pull request #113 from gwenchee/constanttransitionback

Constanttransitionback update

gwenchee committed:
Merge pull request #114 from gwenchee/constanttransitionback

Constanttransitionback update

gwenchee committed:
spent fuel too

gwenchee committed:
so beautiful

gwenchee committed:
Merge pull request #115 from gwenchee/growingtransition

pull

gwenchee committed:
alg

gwenchee committed:
xml update

gwenchee committed:
some changes to plotter

gwenchee committed:
Delete algorithm_performance_tests_constanttransition.py

gwenchee committed:
Delete algorithm_performance_tests_growingtransition.py

gwenchee committed:
Delete algorithm_performance_tests_sinetransition.py

gwenchee committed:
color to prop

pep8speaks commented:
Hello @gwenchee! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2019-07-05 19:50:14 UTC

gwenchee committed:
pep8

gwenchee committed:
pep8

gwenchee committed:
some changes

gwenchee committed:
final parameters used for global paper

FlanFlanagan commented:
These figures look really nice Gwen. 

The changes are huge though. Is it mostly PEP8 stuff?

gwenchee commented:
All the changes in demand_driven_deployment_inst.py are pep 8 issues. The xml files and algorithm_performance_tests_transitions.py are new additions that generate these plots. And I made a few changes in plotter.py to enhance the plots.

FlanFlanagan commented:
Excellent, I'll review those then.

FlanFlanagan commented on d3ploy/plotter.py:
```diff
@@ -204,57 +321,85 @@ def plot_demand_supply_nond3ploy(all_dict, agent_dict, commod, test,
-    if commod.lower() == 'power':
-        ax2.set_ylabel('Power (MW)', fontsize=14)
+    if commod.lower() == "power":
+        ax2.set_ylabel("Power (MW)", fontsize=14)
```
Nitpicky - but this should probably be MWe, since we aren't concerned with the thermal power.

FlanFlanagan commented on tests/performance_tests/algorithm_performance_tests_transitions.py:
```diff
@@ -0,0 +1,150 @@
+    os.remove(file)
+
+ENV = dict(os.environ)
+ENV["PYTHONPATH"] = ".:" + ENV.get("PYTHONPATH", "")
```
Why do you need to update the pythonpath here?

gwenchee replied:
I just tried running the script without that line and it works the same so I removed it. I'm not sure why it is there too

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -31,139 +31,141 @@ class DemandDrivenDeploymentInst(Institution):
-        "institution",
-        alias=['facility_commod', 'facility', 'commod'],
-        uilabel="Facility and Commodities"
+        doc="A map of facilities and each of their corresponding"
```
I'm not sure moving the + to the next line will work.
Does this work for you? 


I'm surprised you don't get this error:

```
TypeError: bad operand type for unary +: 'str'
```

gwenchee replied:
All the changes in this file are pep8 changes when I used `autopep8`

gwenchee replied:
I reverted all pep8 changes

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -31,139 +31,141 @@ class DemandDrivenDeploymentInst(Institution):
+        alias=["facility_pref", "facility", "pref"],
         uilabel="Facility and Preferences",
-        default={}
+        default={},
```
why this comma?

gwenchee replied:
I tested it with and without the comma, and the `pytests` all pass regardless of these commas.

gwenchee replied:
I reverted all pep8 changes

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -31,139 +31,141 @@ class DemandDrivenDeploymentInst(Institution):
+        + " capacities",
+        tooltip="Map of facilities and capacities in the " + "institution",
+        alias=["facility_capacity", "facility", "capacity"],
+        uilabel="Facility and Capacities",
```
why add a comma? This is the final keyword argument item in the function call, no?

gwenchee replied:
I reverted all pep8 changes

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -31,139 +31,141 @@ class DemandDrivenDeploymentInst(Institution):
+        + " output commodities",
+        tooltip="Map of facilities and output commodities in the "
+        + "institution",
+        alias=["facility_commod", "facility", "commod"],
```
What is the point of making this change everywhere? these two things ' and " are equivalent.

gwenchee replied:
I reverted all pep8 changes

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -200,83 +202,114 @@ def enter_notify(self):
+                        self.commodity_dict,
+                        commod,
+                        diff,
+                        time,
```
there should definitely not be a comma here. there's no way this doesn't throw an error.

gwenchee replied:
I reverted all pep8 changes

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -297,15 +330,18 @@ def calc_diff(self, commod, time):
+            demand = self.predict_demand(commod, time) * (
+                1 + self.buffer_dict[commod]
+            )
+        elif self.buffer_type_dict[commod] == "float":
```
This seems like it should be "relative" vs. "absolute" or "percentage" vs "constant". Float is a data type and since percents can be floats, putting these at odds with one another feels fals..

gwenchee replied:
yeah this makes sense. I will address it in another PR. I have made an issue about this: https://github.com/arfc/d3ploy/issues/261

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -297,15 +330,18 @@ def calc_diff(self, commod, time):
         else:
             raise Exception(
-                'You can only choose perc (%) or float (double) for buffer size')
+                "You can only pick perc (%) or float (double) for buffer size"
```
double and float are not synonyms.

gwenchee replied:
I will address this in https://github.com/arfc/d3ploy/issues/261

katyhuff commented on d3ploy/demand_driven_deployment_inst.py:
```diff
@@ -317,43 +353,63 @@ def target(incommod):
+            "poly",
+            "exp_smoothing",
+            "holt_winters",
+            "fft",
```
what is with these commas in the last item of everything?

katyhuff replied:
The closing parenthesis should be here in place of the comma (and everywhere else there is a comma at the end of the set of keyword arguments.)

This is not acceptable and will not run (which would be clear if it were tested):

```
somefunction("variable", 
              "other variable",
              "last variable",
              )
```

This is correct:


```
somefunction("variable", 
             "other variable",
             "last variable")
```

gwenchee replied:
I reverted all pep8 changes

katyhuff reviewed (changes requested):
I'm noticing lots of problems that won't run properly without error. The way multiline strings are being handled is going to throw a TypeError (everywhere) and the way that new commas are being introduced after the last argument in all function calls is also wrong, and is now all over the place. This PR needs tests, because many of these changes will certainly break functionality. I stopped reviewing after looking at the first couple of files. Please apply all requested changes everywhere the error appears (not just where I made comments)

gwenchee committed:
MWe

gwenchee committed:
float is not a double

gwenchee committed:
constant

gwenchee commented:
@katyhuff the comments you made about the commas that were added were added by an `autopep8` software.

gwenchee committed:
globaal edits

gwenchee committed:
global edits

gwenchee committed:
:wq

XMerge branch 'transitionscenarios' of github.com:gwenchee/d3ploy into transitionscenarios

X6xxxx

gwenchee committed:
remove pep8

gwenchee committed:
remove python path update

katyhuff commented on tests/performance_tests/algorithm_performance_tests_transitions.py:
```diff
@@ -0,0 +1,149 @@
+To use this file, in your command line enter:
+python algorithm_performance_tests_transitions.py [transition type]
+There are currently 3 transition types in place:
+constant, growing and sine.
```
```suggestion
constant, growing, and sinunsoidal (sine).
```

katyhuff replied:
@gwenchee I'm trying to trigger a new ci build.

gwenchee committed:
Update tests/performance_tests/algorithm_performance_tests_transitions.py

Co-Authored-By: Katy Huff <katyhuff@users.noreply.github.com>

katyhuff commented on d3ploy/plotter.py:
```diff
@@ -103,96 +154,162 @@ def plot_demand_supply_agent(all_dict, agent_dict, commod, test,
+def plot_demand_supply_nond3ploy(
+    all_dict, agent_dict, commod, test, demand_driven, log_scale, size=6
+):
     """ Plots demand and supply on a curve for a non-driving commodity
```
```suggestion
    """Plots demand and supply on a curve for a non-driving commodity
```

katyhuff replied:
@gwenchee

gwenchee committed:
Update d3ploy/plotter.py

Co-Authored-By: Katy Huff <katyhuff@users.noreply.github.com>

gwenchee committed:
Merge branch 'master' into transitionscenarios

gwenchee committed:
Merge pull request #116 from arfc/master

Update

gwenchee commented:
Awesome! It passes the CI! Time to merge? @FlanFlanagan @robfairh @katyhuff @sonatav2
